<a href="https://colab.research.google.com/github/VighneshS/sentiment_prediction/blob/master/sentiment_prediction.ipynb" target="_blank"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://vighnesh-studies.blogspot.com/2021/04/sentiment-prediction-using-naive-bayes.html" target="_blank">BLOG</a>

# Sentiment Prediction using Naive Bayes Classifier (NBC)
This is a notebook to understand how Naive Bayes Classifier (NBC) works and also how it is useful to classify text based on sentiment.

We will also see how it will be effective against missing data.

## Settings
Training Percentage

In [240]:
training_ratio = 80 / 100
k = 5
most_useful_limit = 20

## Importing the Data
We used the [kaggle dataset](https://storage.googleapis.com/kagglesdsdata/datasets/22169/30047/sentiment%20labelled%20sentences/imdb_labelled.txt?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20210425%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20210425T202010Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=6133706ef10bc2dcd0b58f8398b4d73ab9e9d788de1718b07334df91f6007e1e4ca0b78e3176f95b8250e0c4535ce1633528f4fabffeb7e4124af3ee3f895ac34c03044fca9b23b23c4ddb8fa90d84dfc14869ff4806f03783cafad53b19445b3c3052983fdf1ca4384257eac1bc0a4270d238a1ea89d1289866c7a0ea7ad7c97a76f2e142c148019e39cc5a1295f92650747ac5ea5946b026f7ad6d5d262d4c4a370aee6bc1f5d5b445bb6d93692debe678a79e5e1c1fe3d3e68ea4f2fad3115795d3361e0626e98156fbc7f5967beb7cf0f00e07351d23a00d8677ebb75e3e13b1bfa07762266efabf6f6f9d53206be31b7623cf3614f60f8cf5011cf23def) to get the ground truth of sample IMDB reviews.

In [241]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np  # linear algebra
import pandas as pd  # data processing, CSV file I/O (e.g. pd.read_csv)
from IPython.display import display
import math
from sklearn.model_selection import KFold

In [242]:
# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

# import os
# for dirname, _, filenames in os.walk('/kaggle/input'):
#     for filename in filenames:
#         print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

data = pd.read_csv(
    r"http://storage.googleapis.com/kagglesdsdata/datasets/22169/30047/sentiment%20labelled%20sentences/imdb_labelled.txt?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20210425%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20210425T202010Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=6133706ef10bc2dcd0b58f8398b4d73ab9e9d788de1718b07334df91f6007e1e4ca0b78e3176f95b8250e0c4535ce1633528f4fabffeb7e4124af3ee3f895ac34c03044fca9b23b23c4ddb8fa90d84dfc14869ff4806f03783cafad53b19445b3c3052983fdf1ca4384257eac1bc0a4270d238a1ea89d1289866c7a0ea7ad7c97a76f2e142c148019e39cc5a1295f92650747ac5ea5946b026f7ad6d5d262d4c4a370aee6bc1f5d5b445bb6d93692debe678a79e5e1c1fe3d3e68ea4f2fad3115795d3361e0626e98156fbc7f5967beb7cf0f00e07351d23a00d8677ebb75e3e13b1bfa07762266efabf6f6f9d53206be31b7623cf3614f60f8cf5011cf23def",
    delimiter="\t", header=None, names=["IMDB Review", "Sentiment"])
data = data.sample(frac=1).reset_index(drop=True)

### Split Data
We split the data into train, development and test

In [243]:
train = data[:math.floor(data.shape[0] * training_ratio)]

In [244]:
validation = data[math.floor(data.shape[0] * training_ratio):].sample(frac=1).reset_index(drop=True)
dev, test = np.array_split(validation, 2)

In [245]:
display(train, dev, test)

IMDB Review  Sentiment
0    It has northern humour and positive about the ...          1
1               But it is entertaining, nonetheless.            1
2    The result is a film that just don't look righ...          0
3                              Highly unrecommended.            0
4    Technically, the film is well made with impres...          1
..                                                 ...        ...
593  I am a fan of his ... This movie sucked really...          0
594                     Nothing at all to recommend.            0
595                            It is a true classic.            1
596  It's the one movie that never ceases to intere...          1
597  Don't be afraid of subtitles........ its worth...          1

[598 rows x 2 columns]

IMDB Review  Sentiment
0   This is a bad film, with bad writing, and good...          0
1   Lot of holes in the plot: there's nothing abou...          0
2   Very disappointed and wondered how it could be...          0
3           If you want a real scare rent this one!            1
4   A Lassie movie which should have been "put to ...          0
..                                                ...        ...
70  To be honest with you, this is unbelievable no...          0
71                  The film looks cheap and bland.            0
72  It's a shame to see good actors like Thomerson...          0
73  I could not stand to even watch it for very lo...          0
74                                   I gave it a 10            1

[75 rows x 2 columns]

IMDB Review  Sentiment
75   A very good film indeed, about great and uncon...          1
76   This is truly an art movie--it actually has a ...          1
77   The interplay between Martin and Emilio contai...          1
78   His losing his marbles so early in the proceed...          0
79   In fact, this stinker smells like a direct-to-...          0
..                                                 ...        ...
145                           there is no real plot.            0
146                                It was so BORING!            0
147           It is zillion times away from reality.            0
148        The music in the film is really nice too.            1
149  Almost all of the songs in Cover Girl are old-...          0

[75 rows x 2 columns]

## Generation of Vocabulary list

In [246]:
def split_words(review):
    return review.lower().replace(',', '').replace('"', '').replace('(', '').replace(')', '').replace('\'s',
                                                                                                      '').replace(
        '.',
        '').replace(
        '!', '').replace('-', ' ').replace('/', ' ').split()


def get_word_count(review_data_frame: pd.DataFrame, column_name: str):
    vocab = review_data_frame["IMDB Review"].apply(lambda review: pd.value_counts(
        split_words(review))).count(axis=0).to_frame()
    vocab.columns = [column_name]
    vocab.reset_index(inplace=True)
    vocab = vocab.rename(columns={'index': 'Word'})
    return vocab

## Get Naive Bayes Parameters
Here we have a function to genereate the Naive Bayes Parameters like:

1. Word Frequency
2. P(Word)
3. Positive Sentiment Word Frequency
4. P(Sentiment = Positive)
5. P(Word | Sentiment = Positive)
6. Negative Sentiment Word Frequency
7. P(Sentiment = Negative)
8. P(Word | Sentiment = Negative)

Which are useful in finding:

**P(Sentiment | Sentence (Collection of words)) = P(Sentence | Sentiment) * P(Sentiment) / P(Sentense)**

The P(Sentense) can be approximated to 1 as we are comparing sentiments the value will be cancelled on either sides

In [247]:
def generate_naive_bayes_parameters(data_frame: pd.DataFrame, smoothening: bool):
    naive_bayes_parameters = get_word_count(data_frame, "Word Frequency")
    if smoothening:
        naive_bayes_parameters["Word Frequency"] += 1

    total_words = naive_bayes_parameters["Word Frequency"].sum(axis=0)
    if smoothening:
        total_words += 2

    total_sentiments = data_frame.count(axis=0)['Sentiment']
    if smoothening:
        total_sentiments += 2

    naive_bayes_parameters['P(Word)'] = naive_bayes_parameters["Word Frequency"].div(total_words)

    positive_sentiments = data_frame[data_frame['Sentiment'] == 1]
    positive_vocabulary = get_word_count(positive_sentiments, "Positive Sentiment Word Frequency")
    naive_bayes_parameters = naive_bayes_parameters.merge(positive_vocabulary, how='left', on='Word')
    if smoothening:
        naive_bayes_parameters["Positive Sentiment Word Frequency"] += 1
        naive_bayes_parameters["Positive Sentiment Word Frequency"] = naive_bayes_parameters[
            "Positive Sentiment Word Frequency"].fillna(
            value=1)

    total_positive_words = positive_sentiments.count(axis=0)['Sentiment']
    if smoothening:
        total_positive_words += 2

    probability_of_positive_sentiments = total_positive_words / total_sentiments
    naive_bayes_parameters['P(Sentiment = Positive)'] = probability_of_positive_sentiments

    naive_bayes_parameters['P(Word | Sentiment = Positive)'] = naive_bayes_parameters[
        'Positive Sentiment Word Frequency'].div(
        total_positive_words)

    negative_sentiments = data_frame[data_frame['Sentiment'] == 0]
    negative_vocabulary = get_word_count(negative_sentiments, "Negative Sentiment Word Frequency")
    naive_bayes_parameters = naive_bayes_parameters.merge(negative_vocabulary, how='left', on='Word')
    if smoothening:
        naive_bayes_parameters["Negative Sentiment Word Frequency"] += 1
        naive_bayes_parameters["Negative Sentiment Word Frequency"] = naive_bayes_parameters[
            "Negative Sentiment Word Frequency"].fillna(
            value=1)

    total_negative_words = negative_sentiments.count(axis=0)['Sentiment']
    if smoothening:
        total_negative_words += 2

    probability_of_negative_sentiments = total_negative_words / total_sentiments
    naive_bayes_parameters['P(Sentiment = Negative)'] = probability_of_negative_sentiments

    naive_bayes_parameters['P(Word | Sentiment = Negative)'] = naive_bayes_parameters[
        'Negative Sentiment Word Frequency'].div(
        total_negative_words)

    return naive_bayes_parameters

## To Get the Probabilities

We use this formula to get the probabilities:

**P(Sentiment | Sentence (Collection of words)) = P(Sentence | Sentiment) * P(Sentiment) / P(Sentense)**

The below function will calculate the numerator part and assumes the denominator to be 1 as it will cancel out during
comparison.

For calculating the P(Sentence | Sentiment) we have words in sentences. So, we can write the formula as:

**P(Sentence | Sentiment) = P(Word_1,Word_2,...,Word_n | Sentiment)**

By Naive Bayes Theorem we can write it as:

**P(Word_1,Word_2,...,Word_n | Sentiment) = P(Word_1 | Sentiment).P(Word_2 | Sentiment). ... .P(Word_n | Sentiment)**

In [248]:
def get_probabilities(review: str, naive_bayes_parameters: pd.DataFrame, sentiment: bool, smoothening: bool):
    prob = 1
    column_name = 'P(Word | Sentiment = Positive)' if sentiment else 'P(Word | Sentiment = Negative)'
    individual_prob = 0 if not smoothening else 1 / (
        naive_bayes_parameters['P(Sentiment = Positive)'][0] if sentiment else naive_bayes_parameters[
            'P(Sentiment = Negative)'][0])
    for word in split_words(review):
        if word in naive_bayes_parameters.values:
            individual_prob = naive_bayes_parameters[naive_bayes_parameters['Word'] == word].iloc[0][column_name]
        prob *= 0 if math.isnan(individual_prob) else individual_prob
    return prob * (naive_bayes_parameters['P(Sentiment = Positive)'][0] if sentiment else naive_bayes_parameters[
        'P(Sentiment = Negative)'][0])

In [249]:
def predict_calculate_accuracy(data_frame: pd.DataFrame, naive_bayes_parameters: pd.DataFrame):
    data_frame["P(Sentiment = Positive | Sentence)"] = data_frame["IMDB Review"].apply(
        lambda review: get_probabilities(review, naive_bayes_parameters, True, False))
    data_frame["P(Sentiment = Negative | Sentence)"] = data_frame["IMDB Review"].apply(
        lambda review: get_probabilities(review, naive_bayes_parameters, False, False))
    data_frame["Predicted sentiment"] = data_frame["P(Sentiment = Positive | Sentence)"] > data_frame[
        "P(Sentiment = Negative | Sentence)"]
    accuracy = data_frame.loc[data_frame["Predicted sentiment"] == data_frame["Sentiment"]].count(axis=0)[
                   'Sentiment'] * 100 / data_frame.count(axis=0)['Sentiment']
    print("Accuracy: ", accuracy)
    # print("Wrong Predictions:")
    # display(data_frame.loc[data_frame["Predicted sentiment"] != data_frame["Sentiment"]].reset_index(drop=True))
    return accuracy


## Calculating Accuracy

To calculate accuracy we first divide the training dataset into k parts of train and test the first part of the
set is used to train the dataset with the remaining k-1 test dataset.

We then predict using the Naive bayes parameters that we get from training against the test data.

We then calculate the accuracy by finding (how many data is of correct prediction)/(total number of datasets)

With the parameters having the best accuracy is chosen from this and used for further validation of dev anf test
datasets which we separated in the beginning.

In [250]:
def five_fold_cross_validation(data_frame: pd.DataFrame, smoothening: bool):
    kf = KFold(n_splits=k, shuffle=True)
    train_folds = kf.split(data_frame)
    accuracies = []
    max_accuracy_naive_bayes_parameters = pd.DataFrame()
    for (train_training, train_testing), index in zip(train_folds, range(5)):
        print(f"---------------------------Fold {index + 1}---------------------------------")
        display(train.loc[train_training])
        trained_parameters = generate_naive_bayes_parameters(train.loc[train_training], smoothening)
        accuracy = predict_calculate_accuracy(train.loc[train_testing], trained_parameters)
        accuracies.append(accuracy)
        max_accuracy_naive_bayes_parameters = trained_parameters if max(
            accuracies) == accuracy else max_accuracy_naive_bayes_parameters
        display(trained_parameters)
    return max_accuracy_naive_bayes_parameters


vocabulary = five_fold_cross_validation(train, False)
vocabulary

---------------------------Fold 1---------------------------------


IMDB Review  Sentiment
0    It has northern humour and positive about the ...          1
1               But it is entertaining, nonetheless.            1
2    The result is a film that just don't look righ...          0
3                              Highly unrecommended.            0
4    Technically, the film is well made with impres...          1
..                                                 ...        ...
593  I am a fan of his ... This movie sucked really...          0
594                     Nothing at all to recommend.            0
595                            It is a true classic.            1
596  It's the one movie that never ceases to intere...          1
597  Don't be afraid of subtitles........ its worth...          1

[478 rows x 2 columns]

Accuracy:  56.666666666666664


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0             it             146  0.019120                               78.0   
1     represents               1  0.000131                                1.0   
2            the             250  0.032740                              143.0   
3         humour               2  0.000262                                2.0   
4       northern               1  0.000131                                1.0   
...          ...             ...       ...                                ...   
2400      ceases               1  0.000131                                1.0   
2401       alert               1  0.000131                                1.0   
2402     therapy               1  0.000131                                1.0   
2403   subtitles               1  0.000131                                1.0   
2404    aversion               1  0.000131                                1.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.531381                        0.307087   
1                    0.531381                        0.003937   
2                    0.531381                        0.562992   
3                    0.531381                        0.007874   
4                    0.531381                        0.003937   
...                       ...                             ...   
2400                 0.531381                        0.003937   
2401                 0.531381                        0.003937   
2402                 0.531381                        0.003937   
2403                 0.531381                        0.003937   
2404                 0.531381                        0.003937   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  68.0                 0.468619   
1                                   NaN                 0.468619   
2                                 107.0                 0.468619   
3                                   NaN                 0.468619   
4                                   NaN                 0.468619   
...                                 ...                      ...   
2400                                NaN                 0.468619   
2401                                NaN                 0.468619   
2402                                NaN                 0.468619   
2403                                NaN                 0.468619   
2404                                NaN                 0.468619   

      P(Word | Sentiment = Negative)  
0                           0.303571  
1                                NaN  
2                           0.477679  
3                                NaN  
4                                NaN  
...                              ...  
2400                             NaN  
2401                             NaN  
2402                             NaN  
2403                             NaN  
2404                             NaN  

[2405 rows x 9 columns]

---------------------------Fold 2---------------------------------


IMDB Review  Sentiment
0    It has northern humour and positive about the ...          1
1               But it is entertaining, nonetheless.            1
3                              Highly unrecommended.            0
4    Technically, the film is well made with impres...          1
5    There is no plot here to keep you going in the...          0
..                                                 ...        ...
593  I am a fan of his ... This movie sucked really...          0
594                     Nothing at all to recommend.            0
595                            It is a true classic.            1
596  It's the one movie that never ceases to intere...          1
597  Don't be afraid of subtitles........ its worth...          1

[478 rows x 2 columns]

Accuracy:  70.83333333333333


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0             it             134  0.017503                               70.0   
1     represents               1  0.000131                                1.0   
2            the             241  0.031479                              136.0   
3         humour               2  0.000261                                2.0   
4       northern               1  0.000131                                1.0   
...          ...             ...       ...                                ...   
2388    interest               1  0.000131                                1.0   
2389       alert               1  0.000131                                1.0   
2390     therapy               1  0.000131                                1.0   
2391   subtitles               1  0.000131                                1.0   
2392    aversion               1  0.000131                                1.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.539749                        0.271318   
1                    0.539749                        0.003876   
2                    0.539749                        0.527132   
3                    0.539749                        0.007752   
4                    0.539749                        0.003876   
...                       ...                             ...   
2388                 0.539749                        0.003876   
2389                 0.539749                        0.003876   
2390                 0.539749                        0.003876   
2391                 0.539749                        0.003876   
2392                 0.539749                        0.003876   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  64.0                 0.460251   
1                                   NaN                 0.460251   
2                                 105.0                 0.460251   
3                                   NaN                 0.460251   
4                                   NaN                 0.460251   
...                                 ...                      ...   
2388                                NaN                 0.460251   
2389                                NaN                 0.460251   
2390                                NaN                 0.460251   
2391                                NaN                 0.460251   
2392                                NaN                 0.460251   

      P(Word | Sentiment = Negative)  
0                           0.290909  
1                                NaN  
2                           0.477273  
3                                NaN  
4                                NaN  
...                              ...  
2388                             NaN  
2389                             NaN  
2390                             NaN  
2391                             NaN  
2392                             NaN  

[2393 rows x 9 columns]

---------------------------Fold 3---------------------------------


IMDB Review  Sentiment
2    The result is a film that just don't look righ...          0
3                              Highly unrecommended.            0
4    Technically, the film is well made with impres...          1
5    There is no plot here to keep you going in the...          0
6       Oh yeah, and the storyline was pathetic too.            0
..                                                 ...        ...
586             This movie has a cutting edge to it.            1
589  1/10 - and only because there is no setting fo...          0
590        I'll even say it again – this is torture.            0
596  It's the one movie that never ceases to intere...          1
597  Don't be afraid of subtitles........ its worth...          1

[478 rows x 2 columns]

Accuracy:  60.0


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0         don't              14  0.001910                                3.0   
1          look              13  0.001773                                7.0   
2          just              33  0.004501                               12.0   
3           the             239  0.032601                              135.0   
4          film              68  0.009276                               46.0   
...         ...             ...       ...                                ...   
2272      alert               1  0.000136                                1.0   
2273     afraid               1  0.000136                                1.0   
2274    therapy               1  0.000136                                1.0   
2275  subtitles               1  0.000136                                1.0   
2276   aversion               1  0.000136                                1.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.539749                        0.011628   
1                    0.539749                        0.027132   
2                    0.539749                        0.046512   
3                    0.539749                        0.523256   
4                    0.539749                        0.178295   
...                       ...                             ...   
2272                 0.539749                        0.003876   
2273                 0.539749                        0.003876   
2274                 0.539749                        0.003876   
2275                 0.539749                        0.003876   
2276                 0.539749                        0.003876   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  11.0                 0.460251   
1                                   6.0                 0.460251   
2                                  21.0                 0.460251   
3                                 104.0                 0.460251   
4                                  22.0                 0.460251   
...                                 ...                      ...   
2272                                NaN                 0.460251   
2273                                NaN                 0.460251   
2274                                NaN                 0.460251   
2275                                NaN                 0.460251   
2276                                NaN                 0.460251   

      P(Word | Sentiment = Negative)  
0                           0.050000  
1                           0.027273  
2                           0.095455  
3                           0.472727  
4                           0.100000  
...                              ...  
2272                             NaN  
2273                             NaN  
2274                             NaN  
2275                             NaN  
2276                             NaN  

[2277 rows x 9 columns]

---------------------------Fold 4---------------------------------


IMDB Review  Sentiment
0    It has northern humour and positive about the ...          1
1               But it is entertaining, nonetheless.            1
2    The result is a film that just don't look righ...          0
3                              Highly unrecommended.            0
5    There is no plot here to keep you going in the...          0
..                                                 ...        ...
591  Judith Light is one of my favorite actresses a...          1
592  But it picked up speed and got right to the po...          1
593  I am a fan of his ... This movie sucked really...          0
594                     Nothing at all to recommend.            0
595                            It is a true classic.            1

[479 rows x 2 columns]

Accuracy:  55.46218487394958


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0             it             141  0.018340                               74.0   
1     represents               1  0.000130                                1.0   
2            the             247  0.032128                              134.0   
3         humour               2  0.000260                                2.0   
4       northern               1  0.000130                                1.0   
...          ...             ...       ...                                ...   
2368     setting               1  0.000130                                NaN   
2369    favorite               1  0.000130                                1.0   
2370      judith               1  0.000130                                1.0   
2371      picked               1  0.000130                                1.0   
2372         fan               1  0.000130                                NaN   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.509395                        0.303279   
1                    0.509395                        0.004098   
2                    0.509395                        0.549180   
3                    0.509395                        0.008197   
4                    0.509395                        0.004098   
...                       ...                             ...   
2368                 0.509395                             NaN   
2369                 0.509395                        0.004098   
2370                 0.509395                        0.004098   
2371                 0.509395                        0.004098   
2372                 0.509395                             NaN   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  67.0                 0.490605   
1                                   NaN                 0.490605   
2                                 113.0                 0.490605   
3                                   NaN                 0.490605   
4                                   NaN                 0.490605   
...                                 ...                      ...   
2368                                1.0                 0.490605   
2369                                NaN                 0.490605   
2370                                NaN                 0.490605   
2371                                NaN                 0.490605   
2372                                1.0                 0.490605   

      P(Word | Sentiment = Negative)  
0                           0.285106  
1                                NaN  
2                           0.480851  
3                                NaN  
4                                NaN  
...                              ...  
2368                        0.004255  
2369                             NaN  
2370                             NaN  
2371                             NaN  
2372                        0.004255  

[2373 rows x 9 columns]

---------------------------Fold 5---------------------------------


IMDB Review  Sentiment
0    It has northern humour and positive about the ...          1
1               But it is entertaining, nonetheless.            1
2    The result is a film that just don't look righ...          0
4    Technically, the film is well made with impres...          1
5    There is no plot here to keep you going in the...          0
..                                                 ...        ...
593  I am a fan of his ... This movie sucked really...          0
594                     Nothing at all to recommend.            0
595                            It is a true classic.            1
596  It's the one movie that never ceases to intere...          1
597  Don't be afraid of subtitles........ its worth...          1

[479 rows x 2 columns]

Accuracy:  65.54621848739495


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0             it             131  0.017894                               75.0   
1     represents               1  0.000137                                1.0   
2            the             247  0.033739                              140.0   
3         humour               2  0.000273                                2.0   
4       northern               1  0.000137                                1.0   
...          ...             ...       ...                                ...   
2273      ceases               1  0.000137                                1.0   
2274       alert               1  0.000137                                1.0   
2275     therapy               1  0.000137                                1.0   
2276   subtitles               1  0.000137                                1.0   
2277    aversion               1  0.000137                                1.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.555324                        0.281955   
1                    0.555324                        0.003759   
2                    0.555324                        0.526316   
3                    0.555324                        0.007519   
4                    0.555324                        0.003759   
...                       ...                             ...   
2273                 0.555324                        0.003759   
2274                 0.555324                        0.003759   
2275                 0.555324                        0.003759   
2276                 0.555324                        0.003759   
2277                 0.555324                        0.003759   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  56.0                 0.444676   
1                                   NaN                 0.444676   
2                                 107.0                 0.444676   
3                                   NaN                 0.444676   
4                                   NaN                 0.444676   
...                                 ...                      ...   
2273                                NaN                 0.444676   
2274                                NaN                 0.444676   
2275                                NaN                 0.444676   
2276                                NaN                 0.444676   
2277                                NaN                 0.444676   

      P(Word | Sentiment = Negative)  
0                           0.262911  
1                                NaN  
2                           0.502347  
3                                NaN  
4                                NaN  
...                              ...  
2273                             NaN  
2274                             NaN  
2275                             NaN  
2276                             NaN  
2277                             NaN  

[2278 rows x 9 columns]

Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0             it             134  0.017503                               70.0   
1     represents               1  0.000131                                1.0   
2            the             241  0.031479                              136.0   
3         humour               2  0.000261                                2.0   
4       northern               1  0.000131                                1.0   
...          ...             ...       ...                                ...   
2388    interest               1  0.000131                                1.0   
2389       alert               1  0.000131                                1.0   
2390     therapy               1  0.000131                                1.0   
2391   subtitles               1  0.000131                                1.0   
2392    aversion               1  0.000131                                1.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.539749                        0.271318   
1                    0.539749                        0.003876   
2                    0.539749                        0.527132   
3                    0.539749                        0.007752   
4                    0.539749                        0.003876   
...                       ...                             ...   
2388                 0.539749                        0.003876   
2389                 0.539749                        0.003876   
2390                 0.539749                        0.003876   
2391                 0.539749                        0.003876   
2392                 0.539749                        0.003876   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  64.0                 0.460251   
1                                   NaN                 0.460251   
2                                 105.0                 0.460251   
3                                   NaN                 0.460251   
4                                   NaN                 0.460251   
...                                 ...                      ...   
2388                                NaN                 0.460251   
2389                                NaN                 0.460251   
2390                                NaN                 0.460251   
2391                                NaN                 0.460251   
2392                                NaN                 0.460251   

      P(Word | Sentiment = Negative)  
0                           0.290909  
1                                NaN  
2                           0.477273  
3                                NaN  
4                                NaN  
...                              ...  
2388                             NaN  
2389                             NaN  
2390                             NaN  
2391                             NaN  
2392                             NaN  

[2393 rows x 9 columns]

In [251]:
predict_calculate_accuracy(train, vocabulary)

<ipython-input-249-198896de2224>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_frame["P(Sentiment = Positive | Sentence)"] = data_frame["IMDB Review"].apply(


Accuracy:  93.1438127090301


<ipython-input-249-198896de2224>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_frame["P(Sentiment = Negative | Sentence)"] = data_frame["IMDB Review"].apply(
<ipython-input-249-198896de2224>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_frame["Predicted sentiment"] = data_frame["P(Sentiment = Positive | Sentence)"] > data_frame[


93.1438127090301

In [252]:
predict_calculate_accuracy(dev, vocabulary)

Accuracy:  72.0


72.0

In [253]:
predict_calculate_accuracy(test, vocabulary)

Accuracy:  68.0


68.0

## Most Useful words before Smoothing

This is found by ordering the vocabulary in descending order of **P(Word | Sentiment)** for each negative and positive
sentiments

In [254]:
print("Most Useful Positive sentiment words:")
vocabulary.sort_values("P(Word | Sentiment = Positive)", ascending=False)[:most_useful_limit]

Most Useful Positive sentiment words:


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
2      the             241  0.031479                              136.0   
6      and             171  0.022335                              117.0   
17       a             172  0.022466                              106.0   
75      of             157  0.020507                               93.0   
10      is             153  0.019984                               82.0   
69    this             142  0.018548                               79.0   
0       it             134  0.017503                               70.0   
48      to             120  0.015674                               68.0   
148      i             114  0.014890                               63.0   
37    film              79  0.010319                               52.0   
47      in              94  0.012278                               52.0   
99   movie              81  0.010580                               47.0   
60     was              83  0.010841                               43.0   
44    that              75  0.009796                               38.0   
33    with              49  0.006400                               32.0   
84     for              55  0.007184                               28.0   
12     but              43  0.005617                               28.0   
106    are              43  0.005617                               28.0   
62      as              38  0.004963                               24.0   
34    good              31  0.004049                               23.0   

     P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
2                   0.539749                        0.527132   
6                   0.539749                        0.453488   
17                  0.539749                        0.410853   
75                  0.539749                        0.360465   
10                  0.539749                        0.317829   
69                  0.539749                        0.306202   
0                   0.539749                        0.271318   
48                  0.539749                        0.263566   
148                 0.539749                        0.244186   
37                  0.539749                        0.201550   
47                  0.539749                        0.201550   
99                  0.539749                        0.182171   
60                  0.539749                        0.166667   
44                  0.539749                        0.147287   
33                  0.539749                        0.124031   
84                  0.539749                        0.108527   
12                  0.539749                        0.108527   
106                 0.539749                        0.108527   
62                  0.539749                        0.093023   
34                  0.539749                        0.089147   

     Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
2                                105.0                 0.460251   
6                                 54.0                 0.460251   
17                                66.0                 0.460251   
75                                64.0                 0.460251   
10                                71.0                 0.460251   
69                                63.0                 0.460251   
0                                 64.0                 0.460251   
48                                52.0                 0.460251   
148                               51.0                 0.460251   
37                                27.0                 0.460251   
47                                42.0                 0.460251   
99                                34.0                 0.460251   
60                                40.0                 0.460251   
44                                37.0                 0.460251   
33                                17.0                 0.460251   
84           

In [255]:
print("Most Useful Negative sentiment words:")
vocabulary.sort_values("P(Word | Sentiment = Negative)", ascending=False)[:most_useful_limit]

Most Useful Negative sentiment words:


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
2      the             241  0.031479                              136.0   
10      is             153  0.019984                               82.0   
17       a             172  0.022466                              106.0   
0       it             134  0.017503                               70.0   
75      of             157  0.020507                               93.0   
69    this             142  0.018548                               79.0   
6      and             171  0.022335                              117.0   
48      to             120  0.015674                               68.0   
148      i             114  0.014890                               63.0   
47      in              94  0.012278                               52.0   
60     was              83  0.010841                               43.0   
44    that              75  0.009796                               38.0   
99   movie              81  0.010580                               47.0   
152    bad              30  0.003918                                2.0   
84     for              55  0.007184                               28.0   
37    film              79  0.010319                               52.0   
166     on              40  0.005225                               17.0   
363   just              34  0.004441                               11.0   
209     so              36  0.004702                               14.0   
122    all              37  0.004833                               15.0   

     P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
2                   0.539749                        0.527132   
10                  0.539749                        0.317829   
17                  0.539749                        0.410853   
0                   0.539749                        0.271318   
75                  0.539749                        0.360465   
69                  0.539749                        0.306202   
6                   0.539749                        0.453488   
48                  0.539749                        0.263566   
148                 0.539749                        0.244186   
47                  0.539749                        0.201550   
60                  0.539749                        0.166667   
44                  0.539749                        0.147287   
99                  0.539749                        0.182171   
152                 0.539749                        0.007752   
84                  0.539749                        0.108527   
37                  0.539749                        0.201550   
166                 0.539749                        0.065891   
363                 0.539749                        0.042636   
209                 0.539749                        0.054264   
122                 0.539749                        0.058140   

     Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
2                                105.0                 0.460251   
10                                71.0                 0.460251   
17                                66.0                 0.460251   
0                                 64.0                 0.460251   
75                                64.0                 0.460251   
69                                63.0                 0.460251   
6                                 54.0                 0.460251   
48                                52.0                 0.460251   
148                               51.0                 0.460251   
47                                42.0                 0.460251   
60                                40.0                 0.460251   
44                                37.0                 0.460251   
99                                34.0                 0.460251   
152                               28.0                 0.460251   
84                                27.0                 0.460251   
37           

## Smoothening

Smoothening is done to compensate for unknown words. As all words can't be added to a dictionary and Naive Bayes is
specialized to handle missing words.

Smoothening is done by using the +1 method it is done in the get_naive_bayes_parameters function.

All it does is adding +1 to the following:
1. Word Frequency
2. Positive Sentiment Word Frequency
3. Negative Sentiment Word Frequency

Also, +2 for Number of sentiments as these terms are in the denominator and needs to adhere and compensate for the +1 in
the numerator so that the probability of most occurrence words will be less than 1
1. Total words
2. Total Positive sentiments
3. Total Negative sentiments
4. Total sentiments


In [256]:
vocabulary = five_fold_cross_validation(train, True)
vocabulary

---------------------------Fold 1---------------------------------


IMDB Review  Sentiment  \
0    It has northern humour and positive about the ...          1   
1               But it is entertaining, nonetheless.            1   
2    The result is a film that just don't look righ...          0   
3                              Highly unrecommended.            0   
4    Technically, the film is well made with impres...          1   
..                                                 ...        ...   
593  I am a fan of his ... This movie sucked really...          0   
594                     Nothing at all to recommend.            0   
595                            It is a true classic.            1   
596  It's the one movie that never ceases to intere...          1   
597  Don't be afraid of subtitles........ its worth...          1   

     P(Sentiment = Positive | Sentence)  P(Sentiment = Negative | Sentence)  \
0                          2.471503e-17                        0.000000e+00   
1                          2.276588e-07                        0.000000e+00   
2                          1.782963e-12                        0.000000e+00   
3                          0.000000e+00                        1.901864e-05   
4                          4.910824e-57                        0.000000e+00   
..                                  ...                                 ...   
593                        0.000000e+00                        2.194147e-13   
594                        4.093666e-08                        4.658780e-07   
595                        2.298270e-06                        0.000000e+00   
596                        2.023924e-36                        0.000000e+00   
597                        4.206275e-22                        0.000000e+00   

     Predicted sentiment  
0                   True  
1                   True  
2                   True  
3                  False  
4                   True  
..                   ...  
593                False  
594                False  
595                 True  
596                 True  
597                 True  

[478 rows x 5 columns]

Accuracy:  74.16666666666667


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0             it             132  0.013408                               72.0   
1     represents               2  0.000203                                2.0   
2            the             241  0.024479                              136.0   
3         humour               3  0.000305                                3.0   
4       northern               2  0.000203                                2.0   
...          ...             ...       ...                                ...   
2347      ceases               2  0.000203                                2.0   
2348       alert               2  0.000203                                2.0   
2349     therapy               2  0.000203                                2.0   
2350   subtitles               2  0.000203                                2.0   
2351    aversion               2  0.000203                                2.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.522917                        0.286853   
1                    0.522917                        0.007968   
2                    0.522917                        0.541833   
3                    0.522917                        0.011952   
4                    0.522917                        0.007968   
...                       ...                             ...   
2347                 0.522917                        0.007968   
2348                 0.522917                        0.007968   
2349                 0.522917                        0.007968   
2350                 0.522917                        0.007968   
2351                 0.522917                        0.007968   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  61.0                  0.48125   
1                                   1.0                  0.48125   
2                                 106.0                  0.48125   
3                                   1.0                  0.48125   
4                                   1.0                  0.48125   
...                                 ...                      ...   
2347                                1.0                  0.48125   
2348                                1.0                  0.48125   
2349                                1.0                  0.48125   
2350                                1.0                  0.48125   
2351                                1.0                  0.48125   

      P(Word | Sentiment = Negative)  
0                           0.264069  
1                           0.004329  
2                           0.458874  
3                           0.004329  
4                           0.004329  
...                              ...  
2347                        0.004329  
2348                        0.004329  
2349                        0.004329  
2350                        0.004329  
2351                        0.004329  

[2352 rows x 9 columns]

---------------------------Fold 2---------------------------------


IMDB Review  Sentiment  \
0    It has northern humour and positive about the ...          1   
1               But it is entertaining, nonetheless.            1   
2    The result is a film that just don't look righ...          0   
3                              Highly unrecommended.            0   
4    Technically, the film is well made with impres...          1   
..                                                 ...        ...   
592  But it picked up speed and got right to the po...          1   
593  I am a fan of his ... This movie sucked really...          0   
594                     Nothing at all to recommend.            0   
595                            It is a true classic.            1   
596  It's the one movie that never ceases to intere...          1   

     P(Sentiment = Positive | Sentence)  P(Sentiment = Negative | Sentence)  \
0                          2.471503e-17                        0.000000e+00   
1                          2.276588e-07                        0.000000e+00   
2                          1.782963e-12                        0.000000e+00   
3                          0.000000e+00                        1.901864e-05   
4                          4.910824e-57                        0.000000e+00   
..                                  ...                                 ...   
592                        8.691701e-16                        0.000000e+00   
593                        0.000000e+00                        2.194147e-13   
594                        4.093666e-08                        4.658780e-07   
595                        2.298270e-06                        0.000000e+00   
596                        2.023924e-36                        0.000000e+00   

     Predicted sentiment  
0                   True  
1                   True  
2                   True  
3                  False  
4                   True  
..                   ...  
592                 True  
593                False  
594                False  
595                 True  
596                 True  

[478 rows x 5 columns]

Accuracy:  65.83333333333333


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0             it             137  0.013475                               78.0   
1     represents               2  0.000197                                2.0   
2            the             251  0.024688                              143.0   
3         humour               3  0.000295                                3.0   
4       northern               2  0.000197                                2.0   
...          ...             ...       ...                                ...   
2428       keeps               2  0.000197                                2.0   
2429    decipher               2  0.000197                                2.0   
2430      ceases               2  0.000197                                2.0   
2431    interest               2  0.000197                                2.0   
2432       alert               2  0.000197                                2.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                     0.54375                        0.298851   
1                     0.54375                        0.007663   
2                     0.54375                        0.547893   
3                     0.54375                        0.011494   
4                     0.54375                        0.007663   
...                       ...                             ...   
2428                  0.54375                        0.007663   
2429                  0.54375                        0.007663   
2430                  0.54375                        0.007663   
2431                  0.54375                        0.007663   
2432                  0.54375                        0.007663   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  60.0                 0.460417   
1                                   1.0                 0.460417   
2                                 109.0                 0.460417   
3                                   1.0                 0.460417   
4                                   1.0                 0.460417   
...                                 ...                      ...   
2428                                1.0                 0.460417   
2429                                1.0                 0.460417   
2430                                1.0                 0.460417   
2431                                1.0                 0.460417   
2432                                1.0                 0.460417   

      P(Word | Sentiment = Negative)  
0                           0.271493  
1                           0.004525  
2                           0.493213  
3                           0.004525  
4                           0.004525  
...                              ...  
2428                        0.004525  
2429                        0.004525  
2430                        0.004525  
2431                        0.004525  
2432                        0.004525  

[2433 rows x 9 columns]

---------------------------Fold 3---------------------------------


IMDB Review  Sentiment  \
0    It has northern humour and positive about the ...          1   
1               But it is entertaining, nonetheless.            1   
4    Technically, the film is well made with impres...          1   
6       Oh yeah, and the storyline was pathetic too.            0   
7    Watching washing machine twirling around would...          0   
..                                                 ...        ...   
591  Judith Light is one of my favorite actresses a...          1   
592  But it picked up speed and got right to the po...          1   
593  I am a fan of his ... This movie sucked really...          0   
595                            It is a true classic.            1   
597  Don't be afraid of subtitles........ its worth...          1   

     P(Sentiment = Positive | Sentence)  P(Sentiment = Negative | Sentence)  \
0                          2.471503e-17                        0.000000e+00   
1                          2.276588e-07                        0.000000e+00   
4                          4.910824e-57                        0.000000e+00   
6                          0.000000e+00                        5.478339e-12   
7                          0.000000e+00                        3.107479e-25   
..                                  ...                                 ...   
591                        1.864440e-25                        0.000000e+00   
592                        8.691701e-16                        0.000000e+00   
593                        0.000000e+00                        2.194147e-13   
595                        2.298270e-06                        0.000000e+00   
597                        4.206275e-22                        0.000000e+00   

     Predicted sentiment  
0                   True  
1                   True  
4                   True  
6                  False  
7                  False  
..                   ...  
591                 True  
592                 True  
593                False  
595                 True  
597                 True  

[478 rows x 5 columns]

Accuracy:  70.0


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0             it             135  0.013397                               68.0   
1     represents               2  0.000198                                2.0   
2            the             242  0.024015                              136.0   
3         humour               3  0.000298                                3.0   
4       northern               2  0.000198                                2.0   
...          ...             ...       ...                                ...   
2362      judith               2  0.000198                                2.0   
2363      picked               2  0.000198                                2.0   
2364     therapy               2  0.000198                                2.0   
2365   subtitles               2  0.000198                                2.0   
2366    aversion               2  0.000198                                2.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.539583                        0.262548   
1                    0.539583                        0.007722   
2                    0.539583                        0.525097   
3                    0.539583                        0.011583   
4                    0.539583                        0.007722   
...                       ...                             ...   
2362                 0.539583                        0.007722   
2363                 0.539583                        0.007722   
2364                 0.539583                        0.007722   
2365                 0.539583                        0.007722   
2366                 0.539583                        0.007722   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  68.0                 0.464583   
1                                   1.0                 0.464583   
2                                 107.0                 0.464583   
3                                   1.0                 0.464583   
4                                   1.0                 0.464583   
...                                 ...                      ...   
2362                                1.0                 0.464583   
2363                                1.0                 0.464583   
2364                                1.0                 0.464583   
2365                                1.0                 0.464583   
2366                                1.0                 0.464583   

      P(Word | Sentiment = Negative)  
0                           0.304933  
1                           0.004484  
2                           0.479821  
3                           0.004484  
4                           0.004484  
...                              ...  
2362                        0.004484  
2363                        0.004484  
2364                        0.004484  
2365                        0.004484  
2366                        0.004484  

[2367 rows x 9 columns]

---------------------------Fold 4---------------------------------


IMDB Review  Sentiment  \
0    It has northern humour and positive about the ...          1   
1               But it is entertaining, nonetheless.            1   
2    The result is a film that just don't look righ...          0   
3                              Highly unrecommended.            0   
4    Technically, the film is well made with impres...          1   
..                                                 ...        ...   
592  But it picked up speed and got right to the po...          1   
593  I am a fan of his ... This movie sucked really...          0   
594                     Nothing at all to recommend.            0   
596  It's the one movie that never ceases to intere...          1   
597  Don't be afraid of subtitles........ its worth...          1   

     P(Sentiment = Positive | Sentence)  P(Sentiment = Negative | Sentence)  \
0                          2.471503e-17                        0.000000e+00   
1                          2.276588e-07                        0.000000e+00   
2                          1.782963e-12                        0.000000e+00   
3                          0.000000e+00                        1.901864e-05   
4                          4.910824e-57                        0.000000e+00   
..                                  ...                                 ...   
592                        8.691701e-16                        0.000000e+00   
593                        0.000000e+00                        2.194147e-13   
594                        4.093666e-08                        4.658780e-07   
596                        2.023924e-36                        0.000000e+00   
597                        4.206275e-22                        0.000000e+00   

     Predicted sentiment  
0                   True  
1                   True  
2                   True  
3                  False  
4                   True  
..                   ...  
592                 True  
593                False  
594                False  
596                 True  
597                 True  

[479 rows x 5 columns]

Accuracy:  65.54621848739495


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0             it             140  0.014675                               76.0   
1     represents               2  0.000210                                2.0   
2            the             254  0.026625                              148.0   
3         humour               2  0.000210                                2.0   
4       northern               2  0.000210                                2.0   
...          ...             ...       ...                                ...   
2258       alert               2  0.000210                                2.0   
2259      afraid               2  0.000210                                2.0   
2260     therapy               2  0.000210                                2.0   
2261   subtitles               2  0.000210                                2.0   
2262    aversion               2  0.000210                                2.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.553015                        0.285714   
1                    0.553015                        0.007519   
2                    0.553015                        0.556391   
3                    0.553015                        0.007519   
4                    0.553015                        0.007519   
...                       ...                             ...   
2258                 0.553015                        0.007519   
2259                 0.553015                        0.007519   
2260                 0.553015                        0.007519   
2261                 0.553015                        0.007519   
2262                 0.553015                        0.007519   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  65.0                 0.451143   
1                                   1.0                 0.451143   
2                                 107.0                 0.451143   
3                                   1.0                 0.451143   
4                                   1.0                 0.451143   
...                                 ...                      ...   
2258                                1.0                 0.451143   
2259                                1.0                 0.451143   
2260                                1.0                 0.451143   
2261                                1.0                 0.451143   
2262                                1.0                 0.451143   

      P(Word | Sentiment = Negative)  
0                           0.299539  
1                           0.004608  
2                           0.493088  
3                           0.004608  
4                           0.004608  
...                              ...  
2258                        0.004608  
2259                        0.004608  
2260                        0.004608  
2261                        0.004608  
2262                        0.004608  

[2263 rows x 9 columns]

---------------------------Fold 5---------------------------------


IMDB Review  Sentiment  \
2    The result is a film that just don't look righ...          0   
3                              Highly unrecommended.            0   
5    There is no plot here to keep you going in the...          0   
6       Oh yeah, and the storyline was pathetic too.            0   
9                     This is an extraordinary film.            1   
..                                                 ...        ...   
591  Judith Light is one of my favorite actresses a...          1   
594                     Nothing at all to recommend.            0   
595                            It is a true classic.            1   
596  It's the one movie that never ceases to intere...          1   
597  Don't be afraid of subtitles........ its worth...          1   

     P(Sentiment = Positive | Sentence)  P(Sentiment = Negative | Sentence)  \
2                          1.782963e-12                        0.000000e+00   
3                          0.000000e+00                        1.901864e-05   
5                          0.000000e+00                        1.557680e-17   
6                          0.000000e+00                        5.478339e-12   
9                          4.071717e-05                        1.305057e-05   
..                                  ...                                 ...   
591                        1.864440e-25                        0.000000e+00   
594                        4.093666e-08                        4.658780e-07   
595                        2.298270e-06                        0.000000e+00   
596                        2.023924e-36                        0.000000e+00   
597                        4.206275e-22                        0.000000e+00   

     Predicted sentiment  
2                   True  
3                  False  
5                  False  
6                  False  
9                   True  
..                   ...  
591                 True  
594                False  
595                 True  
596                 True  
597                 True  

[479 rows x 5 columns]

Accuracy:  67.22689075630252


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0         don't              15  0.001540                                4.0   
1          look              15  0.001540                                8.0   
2          just              33  0.003388                               12.0   
3           the             241  0.024741                              130.0   
4          film              79  0.008110                               52.0   
...         ...             ...       ...                                ...   
2308     ceases               2  0.000205                                2.0   
2309      alert               2  0.000205                                2.0   
2310    therapy               2  0.000205                                2.0   
2311  subtitles               2  0.000205                                2.0   
2312   aversion               2  0.000205                                2.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.525988                        0.015810   
1                    0.525988                        0.031621   
2                    0.525988                        0.047431   
3                    0.525988                        0.513834   
4                    0.525988                        0.205534   
...                       ...                             ...   
2308                 0.525988                        0.007905   
2309                 0.525988                        0.007905   
2310                 0.525988                        0.007905   
2311                 0.525988                        0.007905   
2312                 0.525988                        0.007905   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  12.0                  0.47817   
1                                   8.0                  0.47817   
2                                  22.0                  0.47817   
3                                 112.0                  0.47817   
4                                  28.0                  0.47817   
...                                 ...                      ...   
2308                                1.0                  0.47817   
2309                                1.0                  0.47817   
2310                                1.0                  0.47817   
2311                                1.0                  0.47817   
2312                                1.0                  0.47817   

      P(Word | Sentiment = Negative)  
0                           0.052174  
1                           0.034783  
2                           0.095652  
3                           0.486957  
4                           0.121739  
...                              ...  
2308                        0.004348  
2309                        0.004348  
2310                        0.004348  
2311                        0.004348  
2312                        0.004348  

[2313 rows x 9 columns]

Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
0             it             132  0.013408                               72.0   
1     represents               2  0.000203                                2.0   
2            the             241  0.024479                              136.0   
3         humour               3  0.000305                                3.0   
4       northern               2  0.000203                                2.0   
...          ...             ...       ...                                ...   
2347      ceases               2  0.000203                                2.0   
2348       alert               2  0.000203                                2.0   
2349     therapy               2  0.000203                                2.0   
2350   subtitles               2  0.000203                                2.0   
2351    aversion               2  0.000203                                2.0   

      P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
0                    0.522917                        0.286853   
1                    0.522917                        0.007968   
2                    0.522917                        0.541833   
3                    0.522917                        0.011952   
4                    0.522917                        0.007968   
...                       ...                             ...   
2347                 0.522917                        0.007968   
2348                 0.522917                        0.007968   
2349                 0.522917                        0.007968   
2350                 0.522917                        0.007968   
2351                 0.522917                        0.007968   

      Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
0                                  61.0                  0.48125   
1                                   1.0                  0.48125   
2                                 106.0                  0.48125   
3                                   1.0                  0.48125   
4                                   1.0                  0.48125   
...                                 ...                      ...   
2347                                1.0                  0.48125   
2348                                1.0                  0.48125   
2349                                1.0                  0.48125   
2350                                1.0                  0.48125   
2351                                1.0                  0.48125   

      P(Word | Sentiment = Negative)  
0                           0.264069  
1                           0.004329  
2                           0.458874  
3                           0.004329  
4                           0.004329  
...                              ...  
2347                        0.004329  
2348                        0.004329  
2349                        0.004329  
2350                        0.004329  
2351                        0.004329  

[2352 rows x 9 columns]

In [257]:
predict_calculate_accuracy(train, vocabulary)

<ipython-input-249-198896de2224>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_frame["P(Sentiment = Positive | Sentence)"] = data_frame["IMDB Review"].apply(


Accuracy:  93.1438127090301


<ipython-input-249-198896de2224>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_frame["P(Sentiment = Negative | Sentence)"] = data_frame["IMDB Review"].apply(
<ipython-input-249-198896de2224>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_frame["Predicted sentiment"] = data_frame["P(Sentiment = Positive | Sentence)"] > data_frame[


93.1438127090301

In [258]:
predict_calculate_accuracy(dev, vocabulary)

Accuracy:  73.33333333333333


73.33333333333333

In [259]:
predict_calculate_accuracy(test, vocabulary)

Accuracy:  69.33333333333333


69.33333333333333

## Most Useful words after Smoothing

In [260]:
print("Most Useful Positive sentiment words:")
vocabulary.sort_values("P(Word | Sentiment = Positive)", ascending=False)[:most_useful_limit]

Most Useful Positive sentiment words:


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
2      the             241  0.024479                              136.0   
6      and             171  0.017369                              114.0   
19       a             178  0.018080                              105.0   
74      of             152  0.015439                               90.0   
10      is             156  0.015846                               77.0   
68    this             141  0.014322                               75.0   
0       it             132  0.013408                               72.0   
53      to             126  0.012798                               68.0   
108      i             111  0.011275                               61.0   
52      in              90  0.009142                               53.0   
17    film              78  0.007923                               53.0   
97     was              82  0.008329                               45.0   
99   movie              82  0.008329                               43.0   
20    that              81  0.008228                               41.0   
40    with              46  0.004672                               29.0   
12     but              44  0.004469                               29.0   
61      as              47  0.004774                               28.0   
41    good              34  0.003454                               26.0   
83     for              54  0.005485                               26.0   
112    are              41  0.004165                               25.0   

     P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
2                   0.522917                        0.541833   
6                   0.522917                        0.454183   
19                  0.522917                        0.418327   
74                  0.522917                        0.358566   
10                  0.522917                        0.306773   
68                  0.522917                        0.298805   
0                   0.522917                        0.286853   
53                  0.522917                        0.270916   
108                 0.522917                        0.243028   
52                  0.522917                        0.211155   
17                  0.522917                        0.211155   
97                  0.522917                        0.179283   
99                  0.522917                        0.171315   
20                  0.522917                        0.163347   
40                  0.522917                        0.115538   
12                  0.522917                        0.115538   
61                  0.522917                        0.111554   
41                  0.522917                        0.103586   
83                  0.522917                        0.103586   
112                 0.522917                        0.099602   

     Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
2                                106.0                  0.48125   
6                                 58.0                  0.48125   
19                                74.0                  0.48125   
74                                63.0                  0.48125   
10                                80.0                  0.48125   
68                                67.0                  0.48125   
0                                 61.0                  0.48125   
53                                59.0                  0.48125   
108                               51.0                  0.48125   
52                                38.0                  0.48125   
17                                26.0                  0.48125   
97                                38.0                  0.48125   
99                                40.0                  0.48125   
20                                41.0                  0.48125   
40                                18.0                  0.48125   
12           

In [261]:
print("Most Useful Negative sentiment words:")
vocabulary.sort_values("P(Word | Sentiment = Negative)", ascending=False)[:most_useful_limit]

Most Useful Negative sentiment words:


Word  Word Frequency   P(Word)  Positive Sentiment Word Frequency  \
2      the             241  0.024479                              136.0   
10      is             156  0.015846                               77.0   
19       a             178  0.018080                              105.0   
68    this             141  0.014322                               75.0   
74      of             152  0.015439                               90.0   
0       it             132  0.013408                               72.0   
53      to             126  0.012798                               68.0   
6      and             171  0.017369                              114.0   
108      i             111  0.011275                               61.0   
20    that              81  0.008228                               41.0   
99   movie              82  0.008329                               43.0   
52      in              90  0.009142                               53.0   
97     was              82  0.008329                               45.0   
83     for              54  0.005485                               26.0   
169    bad              30  0.003047                                4.0   
17    film              78  0.007923                               53.0   
187     on              41  0.004165                               17.0   
133    all              36  0.003657                               15.0   
16    just              32  0.003250                               12.0   
61      as              47  0.004774                               28.0   

     P(Sentiment = Positive)  P(Word | Sentiment = Positive)  \
2                   0.522917                        0.541833   
10                  0.522917                        0.306773   
19                  0.522917                        0.418327   
68                  0.522917                        0.298805   
74                  0.522917                        0.358566   
0                   0.522917                        0.286853   
53                  0.522917                        0.270916   
6                   0.522917                        0.454183   
108                 0.522917                        0.243028   
20                  0.522917                        0.163347   
99                  0.522917                        0.171315   
52                  0.522917                        0.211155   
97                  0.522917                        0.179283   
83                  0.522917                        0.103586   
169                 0.522917                        0.015936   
17                  0.522917                        0.211155   
187                 0.522917                        0.067729   
133                 0.522917                        0.059761   
16                  0.522917                        0.047809   
61                  0.522917                        0.111554   

     Negative Sentiment Word Frequency  P(Sentiment = Negative)  \
2                                106.0                  0.48125   
10                                80.0                  0.48125   
19                                74.0                  0.48125   
68                                67.0                  0.48125   
74                                63.0                  0.48125   
0                                 61.0                  0.48125   
53                                59.0                  0.48125   
6                                 58.0                  0.48125   
108                               51.0                  0.48125   
20                                41.0                  0.48125   
99                                40.0                  0.48125   
52                                38.0                  0.48125   
97                                38.0                  0.48125   
83                                29.0                  0.48125   
169                               27.0                  0.48125   
17           

## Inference

From the above we can see the effect of smoothening at the time of runtime with accuracy increase of 15%.

Also, from the most useful words we can see 2 things.
1. The most common words are the useful words.
2. The most common words, and some words have higher probability in both positive and negative sentiments.

This shows us that these data need to be removed.

For doing these as future work we can remove stop words from Pythons old NLTK library for stop words.
Also, we can remove the more frequent words like the movie, film as it is both positive and negative which is
logical as it is a movie database...